homeassistant/components/iskra/config_flow.py

```diff
@@ -0,0 +1,264 @@
+    """Test the Modbus connection."""
+    modbus_api = Modbus(
+        ip_address=host, protocol="tcp", port=port, modbus_address=address
+    )
+    try:
+        basic_info = await modbus_api.get_basic_info()
+    except NotAuthorised as e:
+        raise NotAuthorised from e
+    except (DeviceConnectionError, DeviceTimeoutError, InvalidResponseCode) as e:
+        raise CannotConnect from e
+    except Exception as e:
+        _LOGGER.error("Unexpected exception: %s", e)
+        raise UnknownException from e
+
+    return basic_info
+
+
+class IskraConfigFlowFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for iskra."""
+
+    VERSION = 1
+    host: str
+    protocol: str
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle a flow initiated by the user."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            self.host = user_input[CONF_HOST]
+            self.protocol = user_input[CONF_PROTOCOL]
+            if self.protocol == "rest_api":
+                # Check if authentication is required.
+                try:
+                    device_info = await test_rest_api_connection(self.host)
+                except CannotConnect:
+                    errors["base"] = "cannot_connect"
+                except NotAuthorised:
+                    # Proceed to authentication step.
+                    return await self.async_step_authentication()
+                except UnknownException:
+                    errors["base"] = "unknown"
+                    # If the connection was not successful, show an error.
+
+                # If the connection was successful, create the device.
+                if not errors:
+                    return await self._create_entry(
+                        host=self.host,
+                        protocol=self.protocol,
+                        serial=device_info.serial,
+                        model=device_info.model,
+                        port=None,
+                        address=None,
+                        authentication=None,
+                    )
+
+            if self.protocol == "modbus_tcp":
+                # Proceed to modbus step.
+                return await self.async_step_modbus_tcp()
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=STEP_USER_DATA_SCHEMA,
+            errors=errors,
+        )
+
+    async def async_step_authentication(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the authentication step."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            authentication = {
+                "username": user_input[CONF_USERNAME],
+                "password": user_input[CONF_PASSWORD],
+            }
+            try:
+                device_info = await test_rest_api_connection(self.host, authentication)
```

Comment:
```suggestion
            try:
                device_info = await test_rest_api_connection(self.host, user_input)
```
user_input is already shaped like that

Reply:
changed as suggested + to simplify changed _create_entry function